File: thop-go/internal/mcp/handlers.go
package mcp

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/scottgl9/thop/internal/logger"
)

// handleInitialize handles the MCP initialize request
func (s *Server) handleInitialize(ctx context.Context, params json.RawMessage) (interface{}, error) {
	var initParams InitializeParams
	if err := json.Unmarshal(params, &initParams); err != nil {
		return nil, &JSONRPCError{
			Code:    -32602,
			Message: "Invalid params",
			Data:    err.Error(),
		}
	}

	logger.Info("MCP client connected: %s v%s (protocol %s)",
		initParams.ClientInfo.Name,
		initParams.ClientInfo.Version,
		initParams.ProtocolVersion)

	// Return server capabilities
	return InitializeResult{
		ProtocolVersion: MCPVersion,
		Capabilities: ServerCapabilities{
			Tools: &ToolsCapability{
				ListChanged: false,
			},
			Resources: &ResourcesCapability{
				Subscribe:   false,
				ListChanged: false,
			},
			Prompts: &PromptsCapability{
				ListChanged: false,
			},
			Logging: &LoggingCapability{},
		},
		ServerInfo: ServerInfo{
			Name:    "thop-mcp",
			Version: "1.0.0",
		},
	}, nil
}

// handleInitialized handles the initialized notification
func (s *Server) handleInitialized(ctx context.Context, params json.RawMessage) (interface{}, error) {
	logger.Debug("MCP client initialized")
	return nil, nil
}

// handleToolsList handles the tools/list request
func (s *Server) handleToolsList(ctx context.Context, params json.RawMessage) (interface{}, error) {
	tools := []Tool{
		// Session management tools
		{
			Name:        "connect",
			Description: "Connect to an SSH session",
			InputSchema: InputSchema{
				Type: "object",
				Properties: map[string]Property{
					"session": {
						Type:        "string",
						Description: "Name of the session to connect to",
					},
				},
				Required: []string{"session"},
			},
		},
		{
			Name:        "switch",
			Description: "Switch to a different session",
			InputSchema: InputSchema{
				Type: "object",
				Properties: map[string]Property{
					"session": {
						Type:        "string",
						Description: "Name of the session to switch to",
					},
				},
				Required: []string{"session"},
			},
		},
		{
			Name:        "close",
			Description: "Close an SSH session",
			InputSchema: InputSchema{
				Type: "object",
				Properties: map[string]Property{
					"session": {
						Type:        "string",
						Description: "Name of the session to close",
					},
				},
				Required: []string{"session"},
			},
		},
		{
			Name:        "status",
			Description: "Get status of all sessions",
			InputSchema: InputSchema{
				Type:       "object",
				Properties: map[string]Property{},
			},
		},

		// Command execution tool
		{
			Name:        "execute",
			Description: "Execute a command in the active session (optionally in background)",
			InputSchema: InputSchema{
				Type: "object",
				Properties: map[string]Property{
					"command": {
						Type:        "string",
						Description: "Command to execute",
					},
					"session": {
						Type:        "string",
						Description: "Optional: specific session to execute in (uses active session if not specified)",
					},
					"timeout": {
						Type:        "integer",
						Description: "Optional: command timeout in seconds (ignored if background is true)",
						Default:     300,
					},
					"background": {
						Type:        "boolean",
						Description: "Optional: run command in background (default: false)",
						Default:     false,
					},
				},
				Required: []string{"command"},
			},
		},

	}

	return map[string]interface{}{
		"tools": tools,
	}, nil
}

// handleToolCall handles the tools/call request
func (s *Server) handleToolCall(ctx context.Context, params json.RawMessage) (interface{}, error) {
	var callParams ToolCallParams
	if err := json.Unmarshal(params, &callParams); err != nil {
		return nil, &JSONRPCError{
			Code:    -32602,
			Message: "Invalid params",
			Data:    err.Error(),
		}
	}

	logger.Debug("Tool call: %s", callParams.Name)

	// Route to appropriate tool handler
	switch callParams.Name {
	// Session management
	case "connect":
		return s.toolConnect(ctx, callParams.Arguments)
	case "switch":
		return s.toolSwitch(ctx, callParams.Arguments)
	case "close":
		return s.toolClose(ctx, callParams.Arguments)
	case "status":
		return s.toolStatus(ctx, callParams.Arguments)

	// Command execution
	case "execute":
		return s.toolExecute(ctx, callParams.Arguments)

	default:
		return nil, &JSONRPCError{
			Code:    -32601,
			Message: "Unknown tool",
			Data:    fmt.Sprintf("Tool not found: %s", callParams.Name),
		}
	}
}

// handleResourcesList handles the resources/list request
func (s *Server) handleResourcesList(ctx context.Context, params json.RawMessage) (interface{}, error) {
	resources := []Resource{
		{
			URI:         "session://active",
			Name:        "Active Session",
			Description: "Information about the currently active session",
			MimeType:    "application/json",
		},
		{
			URI:         "session://all",
			Name:        "All Sessions",
			Description: "Information about all configured sessions",
			MimeType:    "application/json",
		},
		{
			URI:         "config://thop",
			Name:        "Thop Configuration",
			Description: "Current thop configuration",
			MimeType:    "application/json",
		},
		{
			URI:         "state://thop",
			Name:        "Thop State",
			Description: "Current thop state including session states",
			MimeType:    "application/json",
		},
	}

	return map[string]interface{}{
		"resources": resources,
	}, nil
}

// handleResourceRead handles the resources/read request
func (s *Server) handleResourceRead(ctx context.Context, params json.RawMessage) (interface{}, error) {
	var readParams ResourceReadParams
	if err := json.Unmarshal(params, &readParams); err != nil {
		return nil, &JSONRPCError{
			Code:    -32602,
			Message: "Invalid params",
			Data:    err.Error(),
		}
	}

	var content string
	var err error

	switch readParams.URI {
	case "session://active":
		content, err = s.getActiveSessionResource()
	case "session://all":
		content, err = s.getAllSessionsResource()
	case "config://thop":
		content, err = s.getConfigResource()
	case "state://thop":
		content, err = s.getStateResource()
	default:
		return nil, &JSONRPCError{
			Code:    -32602,
			Message: "Unknown resource URI",
			Data:    readParams.URI,
		}
	}

	if err != nil {
		return nil, &JSONRPCError{
			Code:    -32603,
			Message: "Failed to read resource",
			Data:    err.Error(),
		}
	}

	return ResourceReadResult{
		Contents: []ResourceContent{
			{
				URI:      readParams.URI,
				MimeType: "application/json",
				Text:     content,
			},
		},
	}, nil
}

// handlePromptsList handles the prompts/list request
func (s *Server) handlePromptsList(ctx context.Context, params json.RawMessage) (interface{}, error) {
	prompts := []Prompt{
		{
			Name:        "deploy",
			Description: "Deploy code to a remote server",
			Arguments: []PromptArgument{
				{
					Name:        "server",
					Description: "Target server session name",
					Required:    true,
				},
				{
					Name:        "branch",
					Description: "Git branch to deploy",
					Required:    false,
				},
			},
		},
		{
			Name:        "debug",
			Description: "Debug an issue on a remote server",
			Arguments: []PromptArgument{
				{
					Name:        "server",
					Description: "Server session to debug on",
					Required:    true,
				},
				{
					Name:        "service",
					Description: "Service name to debug",
					Required:    false,
				},
			},
		},
		{
			Name:        "backup",
			Description: "Create a backup of files on a server",
			Arguments: []PromptArgument{
				{
					Name:        "server",
					Description: "Server session to backup from",
					Required:    true,
				},
				{
					Name:        "path",
					Description: "Path to backup",
					Required:    true,
				},
			},
		},
	}

	return map[string]interface{}{
		"prompts": prompts,
	}, nil
}

// handlePromptGet handles the prompts/get request
func (s *Server) handlePromptGet(ctx context.Context, params json.RawMessage) (interface{}, error) {
	var getParams PromptGetParams
	if err := json.Unmarshal(params, &getParams); err != nil {
		return nil, &JSONRPCError{
			Code:    -32602,
			Message: "Invalid params",
			Data:    err.Error(),
		}
	}

	var messages []PromptMessage
	var description string

	switch getParams.Name {
	case "deploy":
		server := getParams.Arguments["server"].(string)
		branch, _ := getParams.Arguments["branch"].(string)
		if branch == "" {
			branch = "main"
		}
		description = fmt.Sprintf("Deploy %s branch to %s server", branch, server)
		messages = []PromptMessage{
			{
				Role: "user",
				Content: Content{
					Type: "text",
					Text: fmt.Sprintf("Please deploy the %s branch to the %s server. "+
						"1. Connect to %s\n"+
						"2. Navigate to the deployment directory\n"+
						"3. Pull the latest changes from %s branch\n"+
						"4. Run any build or deployment scripts\n"+
						"5. Verify the deployment was successful",
						branch, server, server, branch),
				},
			},
		}

	case "debug":
		server := getParams.Arguments["server"].(string)
		service, _ := getParams.Arguments["service"].(string)

		debugText := fmt.Sprintf("Please help me debug an issue on the %s server.", server)
		if service != "" {
			debugText = fmt.Sprintf("Please help me debug the %s service on the %s server.", service, server)
		}

		description = fmt.Sprintf("Debug issue on %s", server)
		messages = []PromptMessage{
			{
				Role: "user",
				Content: Content{
					Type: "text",
					Text: debugText + "\n" +
						"1. Connect to the server\n" +
						"2. Check system resources (CPU, memory, disk)\n" +
						"3. Review relevant logs\n" +
						"4. Identify any errors or issues\n" +
						"5. Suggest fixes or next steps",
				},
			},
		}

	case "backup":
		server := getParams.Arguments["server"].(string)
		path := getParams.Arguments["path"].(string)

		description = fmt.Sprintf("Backup %s from %s", path, server)
		messages = []PromptMessage{
			{
				Role: "user",
				Content: Content{
					Type: "text",
					Text: fmt.Sprintf("Please create a backup of %s on the %s server.\n"+
						"1. Connect to %s\n"+
						"2. Create a timestamped backup of %s\n"+
						"3. Compress the backup\n"+
						"4. Verify the backup was created successfully\n"+
						"5. Report the backup location and size",
						path, server, server, path),
				},
			},
		}

	default:
		return nil, &JSONRPCError{
			Code:    -32602,
			Message: "Unknown prompt",
			Data:    fmt.Sprintf("Prompt not found: %s", getParams.Name),
		}
	}

	return PromptGetResult{
		Description: description,
		Messages:    messages,
	}, nil
}

// handlePing handles ping requests
func (s *Server) handlePing(ctx context.Context, params json.RawMessage) (interface{}, error) {
	return map[string]interface{}{
		"pong": true,
	}, nil
}

// handleCancelled handles cancellation notifications
func (s *Server) handleCancelled(ctx context.Context, params json.RawMessage) (interface{}, error) {
	logger.Debug("Received cancellation notification")
	// TODO: Implement request cancellation
	return nil, nil
}

// handleProgress handles progress notifications
func (s *Server) handleProgress(ctx context.Context, params json.RawMessage) (interface{}, error) {
	var progressParams ProgressParams
	if err := json.Unmarshal(params, &progressParams); err != nil {
		logger.Error("Failed to parse progress params: %v", err)
		return nil, nil
	}

	logger.Debug("Progress update: token=%s progress=%f/%f",
		progressParams.ProgressToken,
		progressParams.Progress,
		progressParams.Total)

	return nil, nil
}

File: thop-rust/src/mcp/handlers.rs
//! MCP request handlers

use serde_json::Value;

use crate::logger;

use super::errors::MCPError;
use super::protocol::{
    InitializeParams, InitializeResult, LoggingCapability, Resource,
    ResourceContent, ResourceReadParams, ResourceReadResult, ResourcesCapability,
    ServerCapabilities, ServerInfo, ToolCallParams, ToolsCapability,
};
use super::server::{Server, MCP_VERSION};
use super::tools;

/// Handle initialize request
pub fn handle_initialize(_server: &mut Server, params: Option<Value>) -> Result<Option<Value>, MCPError> {
    let params_value = params.ok_or_else(|| MCPError::missing_parameter("params"))?;

    let init_params: InitializeParams = serde_json::from_value(params_value)
        .map_err(|e| MCPError::new(super::errors::ErrorCode::InvalidParameter, format!("Invalid params: {}", e)))?;

    logger::info(&format!(
        "MCP client connected: {} v{} (protocol {})",
        init_params.client_info.name,
        init_params.client_info.version,
        init_params.protocol_version
    ));

    let result = InitializeResult {
        protocol_version: MCP_VERSION.to_string(),
        capabilities: ServerCapabilities {
            tools: Some(ToolsCapability { list_changed: false }),
            resources: Some(ResourcesCapability {
                subscribe: false,
                list_changed: false,
            }),
            logging: Some(LoggingCapability {}),
            prompts: None,
            experimental: None,
        },
        server_info: ServerInfo {
            name: "thop-mcp".to_string(),
            version: env!("CARGO_PKG_VERSION").to_string(),
        },
    };

    Ok(Some(serde_json::to_value(result).unwrap()))
}

/// Handle initialized notification
pub fn handle_initialized(_server: &mut Server, _params: Option<Value>) -> Result<Option<Value>, MCPError> {
    logger::debug("MCP client initialized");
    Ok(None)
}

/// Handle tools/list request
pub fn handle_tools_list(_server: &mut Server, _params: Option<Value>) -> Result<Option<Value>, MCPError> {
    let tools = tools::get_tool_definitions();

    let result = serde_json::json!({
        "tools": tools
    });

    Ok(Some(result))
}

/// Handle tools/call request
pub fn handle_tool_call(server: &mut Server, params: Option<Value>) -> Result<Option<Value>, MCPError> {
    let params_value = params.ok_or_else(|| MCPError::missing_parameter("params"))?;

    let call_params: ToolCallParams = serde_json::from_value(params_value)
        .map_err(|e| MCPError::new(super::errors::ErrorCode::InvalidParameter, format!("Invalid params: {}", e)))?;

    logger::debug(&format!("Tool call: {}", call_params.name));

    // Route to appropriate tool handler
    let result = match call_params.name.as_str() {
        "connect" => tools::tool_connect(server, call_params.arguments),
        "switch" => tools::tool_switch(server, call_params.arguments),
        "close" => tools::tool_close(server, call_params.arguments),
        "status" => tools::tool_status(server, call_params.arguments),
        "execute" => tools::tool_execute(server, call_params.arguments),
        _ => {
            return Err(MCPError::new(
                super::errors::ErrorCode::InvalidParameter,
                format!("Unknown tool: {}", call_params.name),
            ));
        }
    };

    Ok(Some(serde_json::to_value(result).unwrap()))
}

/// Handle resources/list request
pub fn handle_resources_list(_server: &mut Server, _params: Option<Value>) -> Result<Option<Value>, MCPError> {
    let resources = vec![
        Resource {
            uri: "session://active".to_string(),
            name: "Active Session".to_string(),
            description: Some("Information about the currently active session".to_string()),
            mime_type: Some("application/json".to_string()),
        },
        Resource {
            uri: "session://all".to_string(),
            name: "All Sessions".to_string(),
            description: Some("Information about all configured sessions".to_string()),
            mime_type: Some("application/json".to_string()),
        },
        Resource {
            uri: "config://thop".to_string(),
            name: "Thop Configuration".to_string(),
            description: Some("Current thop configuration".to_string()),
            mime_type: Some("application/json".to_string()),
        },
        Resource {
            uri: "state://thop".to_string(),
            name: "Thop State".to_string(),
            description: Some("Current thop state including session states".to_string()),
            mime_type: Some("application/json".to_string()),
        },
    ];

    let result = serde_json::json!({
        "resources": resources
    });

    Ok(Some(result))
}

/// Handle resources/read request
pub fn handle_resource_read(server: &mut Server, params: Option<Value>) -> Result<Option<Value>, MCPError> {
    let params_value = params.ok_or_else(|| MCPError::missing_parameter("params"))?;

    let read_params: ResourceReadParams = serde_json::from_value(params_value)
        .map_err(|e| MCPError::new(super::errors::ErrorCode::InvalidParameter, format!("Invalid params: {}", e)))?;

    let content = match read_params.uri.as_str() {
        "session://active" => get_active_session_resource(server)?,
        "session://all" => get_all_sessions_resource(server)?,
        "config://thop" => get_config_resource(server)?,
        "state://thop" => get_state_resource(server)?,
        _ => {
            return Err(MCPError::new(
                super::errors::ErrorCode::InvalidParameter,
                format!("Unknown resource URI: {}", read_params.uri),
            ));
        }
    };

    let result = ResourceReadResult {
        contents: vec![ResourceContent {
            uri: read_params.uri,
            mime_type: Some("application/json".to_string()),
            text: Some(content),
            blob: None,
        }],
    };

    Ok(Some(serde_json::to_value(result).unwrap()))
}

/// Handle ping request
pub fn handle_ping(_server: &mut Server, _params: Option<Value>) -> Result<Option<Value>, MCPError> {
    Ok(Some(serde_json::json!({
        "pong": true
    })))
}

/// Handle cancelled notification
pub fn handle_cancelled(_server: &mut Server, _params: Option<Value>) -> Result<Option<Value>, MCPError> {
    logger::debug("Received cancellation notification");
    Ok(None)
}

/// Handle progress notification
pub fn handle_progress(_server: &mut Server, params: Option<Value>) -> Result<Option<Value>, MCPError> {
    if let Some(params) = params {
        if let Ok(progress) = serde_json::from_value::<super::protocol::ProgressParams>(params) {
            logger::debug(&format!(
                "Progress update: token={} progress={}/{}",
                progress.progress_token,
                progress.progress,
                progress.total.unwrap_or(0.0)
            ));
        }
    }
    Ok(None)
}

// Resource helper functions

fn get_active_session_resource(server: &Server) -> Result<String, MCPError> {
    let session_name = server.sessions.get_active_session_name();
    let session = server.sessions.get_session(session_name)
        .ok_or_else(|| MCPError::no_active_session())?;

    let info = serde_json::json!({
        "name": session_name,
        "type": session.session_type(),
        "connected": session.is_connected(),
        "cwd": session.get_cwd(),
        "environment": session.get_env()
    });

    serde_json::to_string_pretty(&info)
        .map_err(|e| MCPError::new(super::errors::ErrorCode::OperationFailed, format!("Failed to serialize: {}", e)))
}

fn get_all_sessions_resource(server: &Server) -> Result<String, MCPError> {
    let sessions = server.sessions.list_sessions();
    serde_json::to_string_pretty(&sessions)
        .map_err(|e| MCPError::new(super::errors::ErrorCode::OperationFailed, format!("Failed to serialize: {}", e)))
}

fn get_config_resource(server: &Server) -> Result<String, MCPError> {
    serde_json::to_string_pretty(&server.config)
        .map_err(|e| MCPError::new(super::errors::ErrorCode::OperationFailed, format!("Failed to serialize: {}", e)))
}

fn get_state_resource(server: &Server) -> Result<String, MCPError> {
    let active_session = server.state.get_active_session();

    let state_data = serde_json::json!({
        "active_session": active_session,
    });

    serde_json::to_string_pretty(&state_data)
        .map_err(|e| MCPError::new(super::errors::ErrorCode::OperationFailed, format!("Failed to serialize: {}", e)))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::Config;
    use crate::session::Manager as SessionManager;
    use crate::state::Manager as StateManager;

    fn create_test_server() -> Server {
        let config = Config::default();
        let state = StateManager::new(&config.settings.state_file);
        let sessions = SessionManager::new(&config, Some(StateManager::new(&config.settings.state_file)));
        Server::new(config, sessions, state)
    }

    #[test]
    fn test_handle_ping() {
        let mut server = create_test_server();
        let result = handle_ping(&mut server, None).unwrap();
        assert!(result.is_some());
        let value = result.unwrap();
        assert_eq!(value["pong"], true);
    }

    #[test]
    fn test_handle_tools_list() {
        let mut server = create_test_server();
        let result = handle_tools_list(&mut server, None).unwrap();
        assert!(result.is_some());
        let value = result.unwrap();
        assert!(value["tools"].is_array());
    }

    #[test]
    fn test_handle_resources_list() {
        let mut server = create_test_server();
        let result = handle_resources_list(&mut server, None).unwrap();
        assert!(result.is_some());
        let value = result.unwrap();
        assert!(value["resources"].is_array());
    }

    #[test]
    fn test_handle_initialized() {
        let mut server = create_test_server();
        let result = handle_initialized(&mut server, None).unwrap();
        assert!(result.is_none());
    }

    #[test]
    fn test_handle_cancelled() {
        let mut server = create_test_server();
        let result = handle_cancelled(&mut server, None).unwrap();
        assert!(result.is_none());
    }
}
